API functions to control Certificate AltName field

AbramDancy opened the pull request:
Attention @kaleb-himes, ZD 9635

kaleb-himes commented on wolfcrypt/src/asn.c:
```diff
@@ -11250,6 +11250,67 @@ static int wc_SetCert_LoadDer(Cert* cert, const byte* der, word32 derSz)
+	byte hdrBuf[MAX_SEQ_SZ+MAX_OCTET_STR_SZ];
+	int i = 0;
+	int octSize, seqSize, namesSize;
+	byte subjectAltName[] = {0x06, 0x03, 0x55, 0x1D, 0x11};
```
If this is a fixed sequence? If so consider adding to a header asn.h or asn_public.h would be fitting.

kaleb-himes commented on wolfcrypt/src/asn.c:
```diff
@@ -11250,6 +11250,67 @@ static int wc_SetCert_LoadDer(Cert* cert, const byte* der, word32 derSz)
+	
+	// Now we construct the final output into the buffer
+	i = SetSequence(5+octSize+seqSize+namesSize, cert->altNames);		// Initial sequence 0x30, sz
+	XMEMCPY(cert->altNames+i, subjectAltName, 5);			// subjectAltName 0x06, 0x03, 0x55, 0x1D, 0x11
```
magic number 5, consider adding subjectAltName string to header and use XSTRLEN(subjectAltName);

kaleb-himes commented on wolfcrypt/src/asn.c:
```diff
@@ -11250,6 +11250,67 @@ static int wc_SetCert_LoadDer(Cert* cert, const byte* der, word32 derSz)
+	// Now we construct the final output into the buffer
+	i = SetSequence(5+octSize+seqSize+namesSize, cert->altNames);		// Initial sequence 0x30, sz
+	XMEMCPY(cert->altNames+i, subjectAltName, 5);			// subjectAltName 0x06, 0x03, 0x55, 0x1D, 0x11
+	i+=5;
```
magic number 5 incrementing i.

kaleb-himes commented:
@embhorn,

Please followup my review when you have a chance.

AbramDancy commented:
I am unfamiliar with the process here.  Should I make the changes requested and submit a new pull request?

embhorn commented:
> I am unfamiliar with the process here. Should I make the changes requested and submit a new pull request?

After you make changes to address @kaleb-himes comments, you can just push them to the same branch. This PR will remain open.

dgarske commented:
@AbramDancy : Just push changes to your `AbramDancy:Cert-AltName` fork:branch and it will update this PR.

kaleb-himes commented:
@AbramDancy,

If it helps here would be the process:

```
$ cd /your/wolfssl/fork
# <make changes to the code>
$ git status # shows which files have been modified, confirm they look right
$ git diff # A more granular review of the changes
$ git add -A # will add ALL modified files
$ git commit -m "A message describing the changes"
$ git branch # lists the branches in your fork, the one highlighted is the current branch
$ git push origin Name_Of_Branch
```

Let me know if that helps.

- K

embhorn commented on wolfcrypt/src/asn.c:
```diff
@@ -11250,6 +11250,80 @@ static int wc_SetCert_LoadDer(Cert* cert, const byte* der, word32 derSz)
+	/* Append encoded value from name into cert->altNames */
+	byte *buf;
+	int i;
+	
```
Please validate the parameters `cert` and `name`

AbramDancy replied:
input implemented

embhorn commented on wolfcrypt/src/asn.c:
```diff
@@ -11250,6 +11250,80 @@ static int wc_SetCert_LoadDer(Cert* cert, const byte* der, word32 derSz)
+
+int wc_CertAddAltNameStr(Cert *cert, byte *name, cert_altname_type type) {
+	int nameSz;
+	nameSz = XSTRLEN(name);
```
Please validate `name`

embhorn commented on wolfcrypt/src/asn.c:
```diff
@@ -11250,6 +11250,80 @@ static int wc_SetCert_LoadDer(Cert* cert, const byte* der, word32 derSz)
+	int i = 0;
+	int octSize, seqSize, namesSize;
+	byte subjectAltName[] = WC_SUBJECTALTNAME_OID;
+
```
Please validate `cert`

AbramDancy replied:
input implemented

embhorn commented on wolfcrypt/src/asn.c:
```diff
@@ -11250,6 +11250,80 @@ static int wc_SetCert_LoadDer(Cert* cert, const byte* der, word32 derSz)
+	/* hdrBuf will contain:  
+	   Octet header + size  (MAX_OCTET_STR_SZ bytes)
+	   Sequence header + size (MAX_SEQ_SZ bytes) */
+	namesSize = cert->altNamesSz - WC_ALTNAMESOFFSET;
```
Can you validate each of these size values before passing into the next function?

AbramDancy replied:
input implemented (some of this no longer applicable in modified function)

embhorn commented on wolfcrypt/src/asn.c:
```diff
@@ -11250,6 +11250,80 @@ static int wc_SetCert_LoadDer(Cert* cert, const byte* der, word32 derSz)
+	nameSz = XSTRLEN(name);
+	if (type == WC_ALTNAME_RFC822NAME || type == WC_ALTNAME_DNSNAME || 
+		type==WC_ALTNAME_URI) {
+		return wc_CertAddAltName_ex(cert, name, nameSz, type);
```
Better to use a single return. Declare a var, `int ret;` and set it instead of returning. Then `return ret;` at the end.

embhorn commented on wolfcrypt/src/asn.c:
```diff
@@ -11250,6 +11250,80 @@ static int wc_SetCert_LoadDer(Cert* cert, const byte* der, word32 derSz)
+	
+	/* Now we construct the final output into the buffer */
+	/* Initial sequence 0x30, sz */
+	i = SetSequence(WC_SUBJECTALTNAME_OID_SIZE+octSize+seqSize+namesSize, 
```
Very minor, but the use of spaces should be consistent. `WC_SUBJECTALTNAME_OID_SIZE + octSize + seqSize + namesSize,`. There are many places where this applies.

AbramDancy commented:
I just realized that some of my functionality (basically the entire wc_CertEncodeAltName function) is duplicated in the existing function SetAltNames().  I will need to rework things to correct for this -- the header information was getting inserted twice.  This actually simplifies the additions.  

I based my additions on Kaleb's solution in this thread: https://www.wolfssl.com/forums/topic1401-how-to-set-subject-alternative-by-wolfssl-api.html  Note that this solution would have the same issue as my code: all the elements up to the added string value are inserted by the SetAltNames function, which is called (eventually) by the wc_MakeCert API.

AbramDancy commented:
I just realized the source of my error -- the "SetAltNames" function has changed from the build I was working on to the file presently in github.  So Kaleb's solution in the forum would have worked (at the time) and this also explains why my functions worked previously, but now do not.  SetAltNames used to be "dumb" and not add any header information, now it adds the appropriate headers.

AbramDancy commented:
I have attempted to address all listed concerns above, and modified the code to work properly due to the changes in the SetAltNames() function.  I also added test functions in api.c, but I do not have an environment to test that addition and I'm not quite sure if I have set the testing up correctly.

AbramDancy commented:
It appears that the "FlattenAltNames()" function added to the API recently seeks to perform the function I've implemented in a different way.

AbramDancy commented:
I am working on adding the AltNames output to generated CSRs.  This is an update to EncodeCertReq() in asn.c.

embhorn commented on wolfcrypt/src/asn.c:
```diff
@@ -11250,6 +11253,94 @@ static int wc_SetCert_LoadDer(Cert* cert, const byte* der, word32 derSz)
+   WC_ALTNAMESOFFSET is the max header size */
+#define WC_ALTNAMESOFFSET (MAX_SEQ_SZ) 
+
+int wc_CertAddAltName_ex(Cert *cert, byte *name, int nameSz, 
```
Getting compile error:
```
wolfcrypt/src/asn.c:11262:5: error: no previous prototype for ‘wc_CertAddAltName_ex’ [-Werror=missing-prototypes]
 int wc_CertAddAltName_ex(Cert *cert, byte *name, int nameSz,
     ^~~~~~~~~~~~~~~~~~~~
```

AbramDancy replied:
You must have a different compiler settings than me... This function is only intended to be called by the functions below it; would you be satisfied if it were declared "static"?  There are similar functions elsewhere in the build that do not have prototypes.

embhorn commented on wolfcrypt/src/asn.c:
```diff
@@ -11250,6 +11253,94 @@ static int wc_SetCert_LoadDer(Cert* cert, const byte* der, word32 derSz)
+		rv = BAD_FUNC_ARG;
+	} else if (type == WC_ALTNAME_RFC822NAME || type == WC_ALTNAME_DNSNAME || 
+		type==WC_ALTNAME_URI) {
+			nameSz = XSTRLEN(name);
```
Compile error:
```
In file included from ./wolfssl/wolfcrypt/asn.h:29:0,
                 from wolfcrypt/src/asn.c:61:
wolfcrypt/src/asn.c: In function ‘wc_CertAddAltNameStr’:
./wolfssl/wolfcrypt/types.h:428:42: error: pointer targets in passing argument 1 of ‘strlen’ differ in signedness [-Werror=pointer-sign]
         #define XSTRLEN(s1)       strlen((s1))
                                          ^
wolfcrypt/src/asn.c:11298:13: note: in expansion of macro ‘XSTRLEN’
    nameSz = XSTRLEN(name);
             ^~~~~~~
In file included from ./wolfssl/wolfcrypt/types.h:422:0,
                 from ./wolfssl/wolfcrypt/asn.h:29,
                 from wolfcrypt/src/asn.c:61:
/usr/include/string.h:384:15: note: expected ‘const char *’ but argument is of type ‘byte * {aka unsigned char *}’
 extern size_t strlen (const char *__s)
               ^~~~~~
```

AbramDancy replied:
Again, you must have different compiler settings from me to flag this.  I could add "(char *)" in front of the reference, and that should eliminate the error.

embhorn commented on wolfcrypt/src/asn.c:
```diff
@@ -11250,6 +11253,94 @@ static int wc_SetCert_LoadDer(Cert* cert, const byte* der, word32 derSz)
+}
+int wc_CertEncodeAltName(Cert *cert) {
+	/* Encode cert->altNames field by pre-pending the sequence info */
+	int i = 0;
```
Compile error:
```
wolfcrypt/src/asn.c: In function ‘wc_CertEncodeAltName’:
wolfcrypt/src/asn.c:11318:6: error: unused variable ‘i’ [-Werror=unused-variable]
  int i = 0;
      ^
```

AbramDancy replied:
Removed unused variable.

AbramDancy commented on wolfcrypt/src/asn.c:
```diff
@@ -11250,6 +11250,80 @@ static int wc_SetCert_LoadDer(Cert* cert, const byte* der, word32 derSz)
+	                       + MAX_OCTET_STR_SZ + MAX_SEQ_SZ)
+
+int wc_CertAddAltName_ex(Cert *cert, byte *name, int nameSz, 
+						 cert_altname_type type) {
```
Updated asn.c and api.c for style change request.

embhorn commented on tests/api.c:
```diff
@@ -19150,6 +19150,75 @@ static void test_wc_PemPubKeyToDer(void)
+	int oldSize;
+	Cert newCert;
+
+	printf(testFmt,"wc_CertAddAltNameStr()");
```
```
tests/api.c: In function ‘test_wc_CertAddAltName’:
tests/api.c:19162:9: error: ‘testFmt’ undeclared (first use in this function); did you mean ‘resultFmt’?
  printf(testFmt,"wc_CertAddAltNameStr()");
         ^~~~~~~
         resultFmt
tests/api.c:19162:9: note: each undeclared identifier is reported only once for each function it appears in
tests/api.c:19165:34: error: passing argument 2 of ‘wc_CertAddAltNameStr’ discards ‘const’ qualifier from pointer target type [-Werror=discarded-qualifiers]
  ret = wc_CertAddAltNameStr(NULL,"www.altname1.com",WC_ALTNAME_DNSNAME);
                                  ^~~~~~~~~~~~~~~~~~
In file included from tests/api.c:138:0:
./wolfssl/wolfcrypt/asn_public.h:378:21: note: expected ‘byte * {aka unsigned char *}’ but argument is of type ‘const char *’
     WOLFSSL_API int wc_CertAddAltNameStr(Cert *cert, byte *name,
                     ^~~~~~~~~~~~~~~~~~~~
tests/api.c:19175:39: error: passing argument 2 of ‘wc_CertAddAltNameStr’ discards ‘const’ qualifier from pointer target type [-Werror=discarded-qualifiers]
  ret = wc_CertAddAltNameStr(&newCert, "www.altname1.com", WC_ALTNAME_DNSNAME);
                                       ^~~~~~~~~~~~~~~~~~
In file included from tests/api.c:138:0:
./wolfssl/wolfcrypt/asn_public.h:378:21: note: expected ‘byte * {aka unsigned char *}’ but argument is of type ‘const char *’
     WOLFSSL_API int wc_CertAddAltNameStr(Cert *cert, byte *name,
                     ^~~~~~~~~~~~~~~~~~~~
In file included from tests/api.c:145:0:
tests/api.c:19177:22: error: ‘Cert {aka struct Cert}’ has no member named ‘altNameSz’; did you mean ‘altNamesSz’?
  AssertIntGT(newCert.altNameSz,WC_ALTNAMESOFFSET);
                      ^
./tests/unit.h:63:19: note: in definition of macro ‘AssertInt’
     int _x = (int)x;                                                                \
                   ^
tests/api.c:19177:2: note: in expansion of macro ‘AssertIntGT’
  AssertIntGT(newCert.altNameSz,WC_ALTNAMESOFFSET);
  ^~~~~~~~~~~
tests/api.c:19178:20: error: ‘Cert {aka struct Cert}’ has no member named ‘altNameSz’; did you mean ‘altNamesSz’?
  oldSize = newCert.altNameSz;
                    ^~~~~~~~~
                    altNamesSz
tests/api.c:19179:39: error: passing argument 2 of ‘wc_CertAddAltNameStr’ discards ‘const’ qualifier from pointer target type [-Werror=discarded-qualifiers]
  ret = wc_CertAddAltNameStr(&newCert, "www.altname2.com", WC_ALTNAME_DNSNAME);
                                       ^~~~~~~~~~~~~~~~~~
In file included from tests/api.c:138:0:
./wolfssl/wolfcrypt/asn_public.h:378:21: note: expected ‘byte * {aka unsigned char *}’ but argument is of type ‘const char *’
     WOLFSSL_API int wc_CertAddAltNameStr(Cert *cert, byte *name,
                     ^~~~~~~~~~~~~~~~~~~~
In file included from tests/api.c:145:0:
tests/api.c:19181:22: error: ‘Cert {aka struct Cert}’ has no member named ‘altNameSz’; did you mean ‘altNamesSz’?
  AssertIntGT(newCert.altNameSz,oldSize);
                      ^
./tests/unit.h:63:19: note: in definition of macro ‘AssertInt’
     int _x = (int)x;                                                                \
                   ^
tests/api.c:19181:2: note: in expansion of macro ‘AssertIntGT’
  AssertIntGT(newCert.altNameSz,oldSize);
  ^~~~~~~~~~~
tests/api.c:19182:20: error: ‘Cert {aka struct Cert}’ has no member named ‘altNameSz’; did you mean ‘altNamesSz’?
  oldSize = newCert.altNameSz;
                    ^~~~~~~~~
                    altNamesSz
In file included from tests/api.c:145:0:
tests/api.c:19203:22: error: ‘Cert {aka struct Cert}’ has no member named ‘altNameSz’; did you mean ‘altNamesSz’?
  AssertIntGT(newCert.altNameSz,WC_ALTNAMESOFFSET);
                      ^
./tests/unit.h:63:19: note: in definition of macro ‘AssertInt’
     int _x = (int)x;                                                                \
                   ^
tests/api.c:19203:2: note: in expansion of macro ‘AssertIntGT’
  AssertIntGT(newCert.altNameSz,WC_ALTNAMESOFFSET);
  ^~~~~~~~~~~
tests/api.c:19204:20: error: ‘Cert {aka struct Cert}’ has no member named ‘altNameSz’; did you mean ‘altNamesSz’?
  oldSize = newCert.altNameSz;
                    ^~~~~~~~~
                    altNamesSz
In file included from tests/api.c:145:0:
tests/api.c:19207:22: error: ‘Cert {aka struct Cert}’ has no member named ‘altNameSz’; did you mean ‘altNamesSz’?
  AssertIntGT(newCert.altNameSz,oldSize);
                      ^
./tests/unit.h:63:19: note: in definition of macro ‘AssertInt’
     int _x = (int)x;                                                                \
                   ^
tests/api.c:19207:2: note: in expansion of macro ‘AssertIntGT’
  AssertIntGT(newCert.altNameSz,oldSize);
  ^~~~~~~~~~~
tests/api.c:19208:20: error: ‘Cert {aka struct Cert}’ has no member named ‘altNameSz’; did you mean ‘altNamesSz’?
  oldSize = newCert.altNameSz;
                    ^~~~~~~~~
                    altNamesSz
In file included from tests/api.c:145:0:
tests/api.c:19218:22: error: ‘Cert {aka struct Cert}’ has no member named ‘altNameSz’; did you mean ‘altNamesSz’?
  AssertIntLE(newCert.altNameSz,oldSize); /* Encode should shrink buffer */
                      ^
./tests/unit.h:63:19: note: in definition of macro ‘AssertInt’
     int _x = (int)x;                                                                \
                   ^
tests/api.c:19218:2: note: in expansion of macro ‘AssertIntLE’
  AssertIntLE(newCert.altNameSz,oldSize); /* Encode should shrink buffer */
  ^~~~~~~~~~~
tests/api.c:19219:22: error: ‘Cert {aka struct Cert}’ has no member named ‘altNameSz’; did you mean ‘altNamesSz’?
  AssertIntGT(newCert.altNameSz,0);
                      ^
./tests/unit.h:63:19: note: in definition of macro ‘AssertInt’
     int _x = (int)x;                                                                \
                   ^
tests/api.c:19219:2: note: in expansion of macro ‘AssertIntGT’
  AssertIntGT(newCert.altNameSz,0);
  ^~~~~~~~~~~
  CC       examples/server/tests_unit_test-server.o
cc1: all warnings being treated as errors
Makefile:4792: recipe for target 'tests/tests_unit_test-api.o' failed
make[1]: *** [tests/tests_unit_test-api.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: Leaving directory '/home/eric/repos/wolfssl'
Makefile:2925: recipe for target 'all' failed
make: *** [all] Error 2
```

AbramDancy commented:
I am building in Microchip MPLAB X under Windows, an IDE for embedded microcontrollers.  The embedded system does not have any STDIO functions – there is no separate operating system or console running.

From: Eric Blankenhorn <notifications@github.com>
Sent: Monday, March 30, 2020 5:36 PM
To: wolfSSL/wolfssl <wolfssl@noreply.github.com>
Cc: Abram Dancy <dancy@synqor.com>; Mention <mention@noreply.github.com>
Subject: Re: [wolfSSL/wolfssl] API functions to control Certificate AltName field (#2762)


CAUTION: This email originated from outside your organization. Exercise caution when opening attachments or clicking links, especially from unknown senders.

@embhorn requested changes on this pull request.

Hi @AbramDancy<https://github.com/AbramDancy>

Thank you for updating the PR. This is almost ready. What build environment are you using? You should be able to build and run the tests you created. We would be happy to help you get the tests running.

Thanks!
@embhorn<https://github.com/embhorn>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/wolfSSL/wolfssl/pull/2762#pullrequestreview-384249065>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AOLSRCAHVVAZIE2VXL4Y3H3RKEGETANCNFSM4KK4243Q>.

embhorn commented:
> I am building in Microchip MPLAB X under Windows, an IDE for embedded microcontrollers. The embedded system does not have any STDIO functions – there is no separate operating system or console running.

If you are using WIN10, you can easily setup a Windows Subsystem for Linux (WSL) image and run the tests from there.

kaleb-himes commented:
Updated ZD ticket with notes. Closing for now and we will try to get it added in our spare time. Thank you so much for your effort on this!

- K
